### DISCUSSION TITLE
Fail2Ban Started to Ban everyone suddenly

### DISCUSSION THREAD

**Discussion by JaironJunior** [fail2ban/fail2ban#3631] [Q&A]
Hello guys, I would Like to ask for some help.

First as introduction, I would like to say that I'm new to this tool and to the infrastructure of this company i'm currently on, the 2 IT guys that were here, didn't document anything, and I didn't even know this thing existed XD.

We have NGINX reverse proxy, to be used for some sites, wordpress, IIS etc.

Last week, some websites started showing the error 502 for only some external people, and some websites didn't open even for us, so I started digging.

Found out we had that fail2ban on a docker VM, and it was banning multiple IPs at first try, I don't know the reason, didn't find it, but even for us from IT, the ones that did'nt had tried, after trying to access got banned, found out that our external NGINX IP was banned too, so thats why most of the sites wasn't working too, I don't know why it didn't stop for everyone.

I tried looking for logs, files, but I confess that right know it is kind beyond my knowledge to handle, did you guys face anything similar? what logs do you guys need? what can I do for help you find the root of the problem? so far i'm unbanning the IP's and adding them to the Ignore List, but it will get extensive really fast this way :(

**Comment by sebres** [comment]
> I don't know the reason, didn't find it

Take a look at [wiki :: How fail2ban works](https://github.com/fail2ban/fail2ban/wiki/How-fail2ban-works), especially answer to 3rd question is related your issue.

> found out that our external NGINX IP was banned too

Unusual, however it is strictly depending to jails and config (as well as the filters and/or failregexs used).
One method would be to include its address to `ignoreip`, another - to avoid it becomes matched at all or else to find the constellation where it would be to false positive and change it.

> did you guys face anything similar? 

It is impossible because it depends on fail2ban/network/nginx/whatever configuration and can be caused by certain circumstances only, let alone all this is mostly individual.

> what logs do you guys need? 

No one, because one simply cannot blind solve it for you. 
Firstly you need find which jail causes that, e. g. inspect `fail2ban-client status ...` or better grep `fail2ban.log` for IPs that banned mistakenly in your opinion. Hereafter take a look with `fail2ban-regex -o "<msg>" "$logpath" "$filter" | grep "$IP"` which messages would be matched by filter and build an excerpt to check it with `fail2ban-regex -v "$logpath_excerpt" "$filter"` which `failregex` is responsible for that. 
And then either change that regex (to avoid false positives if they are) or resolve the issue on nginx side (if a match is justified and it is just some unfavorable constellation of nginx or behind the proxy).

You can surely provide an excerpt of `fail2ban.log` (with a hint what exactly banned wrong), another excerpt with referenced messages in the logfile/journal (with explanation what/why you consider them as wrong) together with jail/filter configuration, etc, but... 
Better you'd firstly do some home work by yourselves.

> what can I do for help you find the root of the problem?

Nothing (without to know the details), because you alone can find that root. Unless you don't want to give the access for someone and probably pay his work hereafter. Again, the problem is - the details really matter (basically everything may be the issue: wrong or unprecise failregex, some circumstances in nginx but also outside, etc).

**Reply by JaironJunior** [reply]
I'll definitely take a look, may look like I wanted the solved answer, but honestly was more of a desperate move, lost like 3 days until I found this tool and started to have "progress" with it, i'm not a docker/linux guy unfortunately, but seems i'll need to be, just didn't had the time nor the opportunity. So i'll need to firstly study the commands you sent XD, and as soon you read this, you're like "Oh my god, this company is screwed"

I'll do some home work, so if you would sum up the files that are the "rules" lets say, that I would need to look in order to consider what to ban and what not, what would then be? I just looked into the action ones, filters, jail.conf, jail.local, but honestly didn't find something related to it, and since it's banning people that used to access our websites at first try, I find it odd that it started all of sudden.

Started lookin into the fail2ban/var/log/fail2ban.log and I've seen this message, should I disconsider? and If I indeed needed to change it, would it be too difficult? We just have this one jail, don't know why they named this way
![image](https://github.com/fail2ban/fail2ban/assets/151084383/983db38c-e375-4f50-a407-4596be7f628a)

Here is an example of an IP that kept getting blocked and used to have no problem as far as I know.
```
2023-11-08 16:50:52,474 fail2ban.utils          [1]: ERROR   7fd92b27bcc0 -- exec: iptables -I DOCKER-USER -m string --algo bm --string 'X-Forwarded-For: 181.192.82.200' -j DROP
bash /data/action.d/whatsapp_notif.sh -b 181.192.82.200 -r "forceful browsing on 192.168.xxx.xxx (npm-general-forceful-browsing)"
2023-11-08 16:50:52,474 fail2ban.utils          [1]: ERROR   7fd92b27bcc0 -- timed out after 60 seconds.
2023-11-08 16:50:52,674 fail2ban.utils          [1]: ERROR   iptables -I DOCKER-USER -m string --algo bm --string 'X-Forwarded-For: 181.192.82.200' -j DROP
bash /data/action.d/whatsapp_notif.sh -b 181.192.82.200 -r "forceful browsing on 192.168.xxx.xxx (npm-general-forceful-browsing)" -- failed with [Errno 3] No such process
```
That ^ whatsapp_notif thing, is not working apparently, I'll need to take a look into it as well, but I've been here for like 1month and a half, so, theres lots of things to look into...
```
2023-11-08 16:51:52,899 fail2ban.actions        [1]: ERROR   Failed to execute ban jail 'npm-general-forceful-browsing' action 'action-ban-docker-forceful-browsing' info 'ActionInfo({'ip': '181.192.82.200', 'family': 'inet4', 'fid': <function Actions.ActionInfo.<lambda> at 0x7fd92af08700>, 'F-*': {'matches': [['[', '08/Nov/2023:04:16:24 -0300', '] - 401 401 - GET https hereis.mywebsite.com.br "/wfrmLogin.aspx/ValidaToken" [Client 181.192.82.200] [Length 1293] [Gzip -] [Sent-to 192.1xx.xxx.xxx(VM with IIS from website).5.81] "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36" "https://hereis.mywebsite.com.br/app/resultados"']], 'failures': 2, 'ip4': '181.192.82.200', 'container': '192.1xx.xxx.xxx(VM with IIS from website).5.81', 'useragent': '"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36"'}, 'raw-ticket': <function Actions.ActionInfo.<lambda> at 0x7fd92af08dc0>})': Error banning 181.192.82.200
```
```
[npm-general-forceful-browsing] Restore Ban 181.192.82.200

[npm-general-forceful-browsing] Unban 181.192.82.200

 Found 181.192.82.200 - 2023-11-09 17:19:00
 Found 181.192.82.200 - 2023-11-09 17:19:00
 Found 181.192.82.200 - 2023-11-09 17:19:01
 Found 181.192.82.200 - 2023-11-09 17:19:01
 Found 181.192.82.200 - 2023-11-09 17:19:01
 fail2ban.actions        [1]: NOTICE  [npm-general-forceful-browsing] Ban 181.192.82.200
 
Found 181.192.82.200 - 2023-11-17 08:17:04
Found 181.192.82.200, bad - 2023-11-17 08:17:04, 3 # -> 5.0
Ban 181.192.82.200
```

that IP is on my "jail.local"
![image](https://github.com/fail2ban/fail2ban/assets/151084383/6a32956f-a8ee-486d-8958-8e18f01b9799)
and I already used "fail2ban-client set npm-general-forceful-browsing unbanip 181.192.82.200"

just used the command above again, just in case, i'll keep an eye on the log file, and i'll study the commands you've sent

So far, Thank you!!!!!!
Really appreciated

**Reply by sebres** [reply]
> I've seen this message, should I disconsider?

Well By iptables a user-defined chain name has a maximum length of 31 characters, together with fail2ban prefix it'd get `f2b-npm-general-forceful-browsing`, that would be 33 chars long. Just the action seems to be a custom action - it doesn't use chains (or at least I don't see it in your excerpts)...

>  and If I indeed needed to change it, would it be too difficult?

One'd need to rename the section in jail.local (or jail.conf if used instead jail.local), for example:
```diff
- [npm-general-forceful-browsing]
+ [npm-gen-frcfl-browsing]
...
```
(don't forget to restart fail2ban hereafter)

> That ^ whatsapp_notif thing, is not working apparently

Yes it seems to reach a timeout (default 60 sec). Such notifying things running synchronously are not recommended at all - either it should be rewritten asynchronously or/and it must send statistic for several bans as summary (and to get 1 message for ever ban is a bit too heavy anyway).

But also the rest of action looks weird - it is not IP based, but rather string based - blocking packets containing string `X-Forwarded-For: <ip>`, what is correct related to real address (because if I correct understand the real IP is the IP of the nginx behind the proxying side and and `X-Forwarded-For` or `X-Real-IP` only part of the header), but not fully correct in sense of whole handling: the header may stretch across 2 packets, it may be encrypted (https/http2/http3/etc), etc. 
Worse thing will be multiplexing (if several requests combined to single buffer reaching iptables as a single packet in one piece, which goes dropped, so reject multiple request at once).
Better would be to use [nginx-block-map](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/nginx-block-map.conf) action (that would ban directly in nginx), see https://github.com/fail2ban/fail2ban/issues/2258#issuecomment-435105548 for similar constellation.

> that IP is on my "jail.local"

Indeed, it will also be ignored by `181.192.0.0/16` subnet.
Did you reloaded fail2ban or jail after adding that entries to `ignoreip`? `?sudo? fail2ban-client reload`
If not - this is the issue. If yes - we'd need the dump of config stream to investigate deeper `?sudo? fail2ban-client -d | grep npm-general-forceful-browsing`.
